spec/Section 5 -- Validation.md

```diff
@@ -1431,6 +1452,103 @@ arguments, an input object may have required fields. An input field is required
 if it has a non-null type and does not have a default value. Otherwise, the
 input object field is optional.
 
+### OneOf Input Objects Have Exactly One Field
+
+**Formal Specification**
+
+- For each {operation} in {document}:
+  - Let {oneofInputObjects} be all OneOf Input Objects transitively included in
+    the {operation}.
+  - For each {oneofInputObject} in {oneofInputObjects}:
+    - Let {fields} be the fields provided by {oneofInputObject}.
+    - {fields} must contain exactly one entry.
+    - Let {field} be the sole entry in {fields}.
+    - Let {value} be the value of {field}.
+    - {value} must not be the {null} literal.
+    - If {value} is a variable:
+      - Let {variableName} be the name of {variable}.
+      - Let {variableDefinition} be the {VariableDefinition} named
+        {variableName} defined within {operation}.
+      - Let {variableType} be the expected type of {variableDefinition}.
+      - {variableType} must be a non-null type.
```

Comment:
I wonder if these lines relating to variable usage pre-empt the discussion around https://github.com/graphql/graphql-spec/pull/1059 and should be pulled from this spec change (simplifying it).

Variables must be only of the allowed type, but it seems that we should specify what that entails for all variables and types only in one place, i.e. the separate rule.

So if we currently allow variables of nullable types to be used in non-null positions and throw a field error at runtime -- which we do -we should continue to do so irrespective of isOneOf, and if/when we make the change there, that should be done in a way that covers isOneOf as well.

Encountered this while attempting to rebase

Reply:
https://github.com/graphql/graphql-js/pull/3813

Reply:
There are two arguments against this:

1. The benefit of the current relaxed version of `VariablesInAllowedPositions` is that you can use a variable for an argument, not supply the value, and get the default. But with OneOf, default values are not allowed for any fields, so treating this as the same as the general case would only be sensical if (a) we adopt the strict version of the general rule (b) we can convince ourselves that there is a real value in consistency almost for consistency's case.
2.  We can only consider this to be a specific case of the general rule if the `@oneof` directive is held to transform all of the input object's field types into non-nullable (but still not required) types. Then, these become non-nullable positions. There is a certain ambiguity as to whether the field types themselves are nullable or not. By syntax, we want to make sure older clients can leave them out, and so we define them to be nullable. But for clients aware of `@oneof`, presumably we are ok to define them as non-nullable, with the caveat that there would have to be a change to the `IsRequiredInputField` algorithm. Currently, an input object field is required if is of a non-null type and does not have a default value. This would have to be changed to have an additional condition, that the parent input object does not have isOneOf to be true. Note that `IsRequiredInputField` is part of graphql-js as a utility, and referenced many times in the spec, but does form a formal agorithm.

Reply:
> I wonder if these lines relating to variable usage pre-empt the discussion around https://github.com/graphql/graphql-spec/pull/1059 and should be pulled from this spec change (simplifying it).

I don't think so? Technically all fields on a oneof are nullable, but you must specify one and it must be non-null, so this seems a very straightforward way to require that when it comes to a variable? #1059 handles non-null positions, but this is a nullable position according to the type system.

Reply:
> and it must be non-null

So it's a nullable type only because we want introspection to say that it's nullable because currently that's the only way of making something optional. But a null cannot be supplied, so in a sense it's by definition "a non-nullable position."

So we would then have to introduce the concept of non-nullable positions that occur when (1) the type for the position is non-nullable or (2) the containing type is oneOf, and then the general rule about matching nullable and non-nullable would have to depend on this new "position" concept rather than the type itself.

As I type this, I can see that this additional layer is a bad idea, and I appreciate the compromise that you have ended up with.

On the other hand, in GraphQL 2.0 / TreeQL, we should definitely separate optionality and nullability, and remember to change oneOf to be better defined. (It really shouldn't be the case that you cannot use null at nullable position.)